gateway_android.go

```diff
@@ -0,0 +1,50 @@
+package gateway
+
+import (
+	"net"
+	"os/exec"
+)
+
+// some android versions has no default gateway ip. or route table. it can have multiple '0.0.0.0' gates with same metrics at the same time.
+
+func DiscoverGateway() (ip net.IP, err error) {
+	ip, err = discoverGatewayUsingRoute()
+	if err != nil {
+		ip, err = discoverGatewayUsingIp()
+		if err != nil {
+			ip, err = discoverGatewayUsingPing()
+		}
+	}
+	return
+}
+
+func discoverGatewayUsingIp() (net.IP, error) {
+	routeCmd := exec.Command("/system/bin/ip", "route", "show")
+	output, err := routeCmd.CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	return parseLinuxIPRoute(output)
+}
+
+func discoverGatewayUsingRoute() (net.IP, error) {
+	routeCmd := exec.Command("/system/bin/route", "-n")
+	output, err := routeCmd.CombinedOutput()
+	if err != nil {
+		return nil, err
+	}
+
+	return parseLinuxRoute(output)
+}
+
+// unsucessful ping discover can take up to 1 sec
+func discoverGatewayUsingPing() (net.IP, error) {
+	routeCmd := exec.Command("/system/bin/ping", "-n", "-c", "1", "-t", "0", "-W", "1", "198.41.0.4") // a.root-servers.net
```

Comment:
Are you sure this should be "-t 0", and not "-t 1"?

If I try (on a random Linux box I happen to have handy, I get):

$ ping -n -c 1 -t 0  -W 1 198.41.0.4
ping: can't set unicast time-to-live: Invalid argument

But if I try with -t 1, it works

$ ping -n -c 1 -t 1  -W 1 198.41.0.4
PING 198.41.0.4 (198.41.0.4) 56(84) bytes of data.
From 192.168.1.1: icmp_seq=1 Time to live exceeded

--- 198.41.0.4 ping statistics ---
1 packets transmitted, 0 received, +1 errors, 100% packet loss, time 0ms

$